### PR TITLE
CI: generic runtime tests

### PR DESCRIPTION
**Pull request by aparcar**
Requested via https://github.com/openwrt/packages/issues/13589

### CODE REVIEW

**Comment by aparcar** [comment]
@jefferyto Here we face the first problem. The `vim` package contains a tool called `xxd` which is an entirely different version:

```
Testing package xxd in version 8.2 from vim
Installing xxd (8.2-2) to root...
Configuring xxd.
Use generic test
Test executable /usr/bin/xxd
```

**Comment by jefferyto** [comment]
I don't want to comment too much on the immediate issue with `xxd` because I'm not crazy about this PR as it stands. I think doing everything inside the container is too inflexible and you have no access to the package Makefile.

I suggest something like this (it doesn't need to be all done in one PR), for each package/subpackage:

* Checking the package Makefile, if the package has opted into tests (as mentioned in https://github.com/openwrt/packages/issues/13589#issuecomment-717119578)
  * If the package does not have a custom test script, output the default test script to something like `custom_test.sh` or `custom_test_<subpackage-name>.sh`
  * (optional) If the package has a customized command line argument (instead of `--version -version ...`), output the default test script with the custom command line argument into `custom_test.sh` / `custom_test_<subpackage-name>.sh` (perhaps there are other variables in the default test script that can be customized in a similar manner)
  * If the package has a custom test script, output it into `custom_test.sh` / `custom_test_<subpackage-name>.sh`
* Inside the container
  * If there is a `custom_test.sh` / `custom_test_<subpackage-name>.sh`, run it
  * Otherwise if there is a `test.sh`, run that

This way the content of `test.sh` can even be completely defined inside the Makefile (similar to config options, they can be defined inside the Makefile or `source`d from a separate file).

**Comment by champtar** [comment]
We expect contributor to be able to fix CI failure, so we need to be able to replicate what the CI does as easily as possible
=> `make package/aaa/test` would be really nice
And also CI should be green for working packages, it's nice to test more things but if we end up ignoring 90% of the times is it really better

**Comment by aparcar** [comment]
Sorry this slipped through my notifications.

> everything inside the container is too inflexible

Please elaborate what parts are too inflexible.

> you have no access to the package Makefile

The `test.sh` script is placed literally next to the Makefile and therefore accessible from the container. In fact the entire `packages.git` is mounted within the Docker container.

We could use Makefile `define` to *define* tests rather than having separate files, not sure if that's a big advantage.

>  => make package/aaa/test would be really nice

How do you envision this to be implemented? Your host system is unlikely to support all these architectures, likely not even x86-musl. Then you need a working `opkg` installing dependencies. You can make it work, but that's plenty of work. One could automatically spin up a dozen Docker containers, but then again that's what the CI is for. My lazy approach would be that people just draft PRs and push until the CI is happy. I don't expect anyone to test their PRs for 10 archs.

File: .github/workflows/entrypoint.sh [hunk 1]
```diff
@@ -32,7 +32,34 @@ for PKG in /ci/*.ipk; do
...
+				done
+
+			fi
+		done
+		if [ "$FOUND_EXE" -eq 1 ]; then
+			if [ "$SUCCESS" -eq 0 ]; then
+				echo "Generic test failed"
+				exit 1
+			else
+				echo "Test succesful or"
```
Comment (BKPepe) [inline .github/workflows/entrypoint.sh:58]: Shouldn't this be without or?

```suggestion
				echo "Test successful"
```

File: .github/workflows/entrypoint.sh [hunk 2]
```diff
@@ -32,7 +32,34 @@ for PKG in /ci/*.ipk; do
...
+			fi
+		else
+			echo "No executeable found in package $PKG_NAME"
```
Comment (BKPepe) [inline .github/workflows/entrypoint.sh:61]:
```suggestion
			echo "No executable found in package $PKG_NAME"
```

**Comment by BKPepe** [comment]
Tested with lighttpd. Seems OK to me!  More tested packages will follow soon.

**Comment by aparcar** [comment]
I'd like to get more input from @champtar and @jefferyto before continuing.

**Comment by BKPepe** [comment]
> * (optional) If the package has a customized command line argument (instead of `--version -version ...`), output the default test script with the custom command line argument into `custom_test.sh` / `custom_test_<subpackage-name>.sh` (perhaps there are other variables in the default test script that can be customized in a similar manner)

I agree on this one. We should have something like this. 
There might be some packages, which does not have e.g. `--version`, `-version` and it that case check fails and in the logs, there is: generic test failed. In that case, there could be fallback to custom test or make it even complicated: 

1. try to check for `version`
2. try to show `help`
3. try to run app without any parameters to check if it runs and that it does not segfault or does not show any errors
4. have custom script if method 1.,2. fails (they are not there).

Because I doubt that anyone wants to create his/her `test.sh` or  ship the generic `test.sh` with each package folder, so having something by default :+1: 

TL;DR: I agree with @jefferyto's suggestions, but one difference that all packages will be using tests even though could be complicated for libraries.

**Comment by BKPepe** [comment]
It looks like several members are really depend on run testing packages inside CI/CD, but most of those packages does not have test.sh file at all and it is not run tested in the end. We should proceed with this PR.

Any thoughts, @1715173329 , @neheb, @robimarko ?

File: .github/workflows/entrypoint.sh [hunk 3]
```diff
@@ -32,7 +32,34 @@ for PKG in /ci/*.ipk; do
 			exit 1
 		fi
 	else
-		echo "No test.sh script available"
+		echo "Use generic test"
+	        PKG_FILES=$(opkg files "$PKG_NAME" | grep -E "/bin|/sbin|/usr/bin|/usr/sbin")
```
Comment (GeorgeSapkin) [inline .github/workflows/entrypoint.sh:36]:
```suggestion
			PKG_FILES=$(opkg files "$PKG_NAME" | grep -E "/bin|/sbin|/usr/bin|/usr/sbin")
```

File: .github/workflows/entrypoint.sh [hunk 4]
```diff
@@ -32,7 +32,34 @@ for PKG in /ci/*.ipk; do
...
+			if [ -x "$FILE" ] && [ -f "$FILE" ]; then
+				FOUND_EXE=1
+				echo "Test executable $FILE"
+				for V in --version -version version -v -V; do
+					echo "Trying $FILE $V"
+					if "$FILE" "$V" 2>&1 | grep "$PKG_VERSION"; then
+						SUCCESS=1
+						break
+					fi
+				done
```
Comment (GeorgeSapkin) [inline .github/workflows/entrypoint.sh:49]: I think ensuring the main binary can display help should be enough for a success already, as a last resort after the version checks. This at least checks that binaries are valid and have the linked dependencies installed.

Also I think it makes sense to do `grep -F` if that's not the default.

Comment (wehagy) [reply .github/workflows/entrypoint.sh:49]: But how we can test this? Because not all binaries return `exit 0`, maybe `grep` the binary name?

**Review by wehagy — COMMENTED** [review]
Also is better move to https://github.com/openwrt/actions-shared-workflows/blob/main/.github/dockerfiles_feeds/entrypoint.sh?

File: .github/workflows/entrypoint.sh (hunk 3 above)
Comment (wehagy) [reply .github/workflows/entrypoint.sh:36]: Need `apk info -L` support

File: .github/workflows/entrypoint.sh (hunk 4 above)
Comment (wehagy) [reply .github/workflows/entrypoint.sh:49]: > I think ensuring the main binary can display help should be enough for a success already, as a last resort after the version checks. This at least checks that binaries are valid and have the linked dependencies installed

The problem is how? Because not all packages use `exit 0` as a return. Maybe we can use `grep` to check the binary name itself?

Comment (GeorgeSapkin) [reply .github/workflows/entrypoint.sh:49]: I meant on top of `--version -version version -v -V` do something like `--help`, etc. Worst case scenario somebody's PR fails and they add a proper test.

Might be a good idea to also do an `ldd` (if it's a binary) and check if any linked dependencies are missing. And do this even if packages have their own tests.

**Comment by GeorgeSapkin** [comment]
I ported this PR to https://github.com/openwrt/actions-shared-workflows/pull/60 since this logic no-longer lives here. Can this be closed?
